bpo-28685: Optimize sorted() list.sort() with type-specialized comparisons

embg opened the pull request:
## Description of the optimization (see also [this poster](https://github.com/embg/python-fastsort-benchmark/blob/master/poster.pdf))
The idea is simple: in practice, it's very uncommon to sort type-heterogeneous lists. This is because lists in general tend to be used in a homogeneous way (if you're iterating and the type is changing, your code may break, depending on what you're doing), and because comparison is often not defined in the heterogeneous context ("apples and oranges"). 

So, instead of checking types during every single compare in the sort (dynamic dispatch), we can simply iterate once in a pre-sort check and see if the list is type-homogeneous. If it is, we can replace `PyObject_RichCompareBool` with whatever compare function would have ended up being dispatched for that type. Since this check is cheap and very unlikely to fail, and checking types every time we compare is expensive, this is a reasonable optimization to consider.

This is, however, only the beginning of what's possible. Namely, there are many safety checks that have to be performed during every compare in the common cases (string, int, float, tuple) that one encounters in practice. For example, character width has to be checked for both strings every time two strings are compared. Since these checks almost never fail in practice (because, e.g., non-latin strings are uncommon in practice, etc.), we can move them out of the comparison function and into the pre-sort check, as well. We then write special-case compare functions (I implemented one for each of the four types mentioned above) that are selected iff. the assumptions necessary to use them are verified for each list element. 

## Benchmarks
I considered two sets of benchmarks: one organized by type (random lists of that type), and one organized by structure. Full benchmark scripts can be found [here](https://github.com/embg/python-fastsort-benchmark). The results are below (standard deviations were less than 0.3% of the mean for all measurements):

### By type

Type | Percent improvement on random lists of [type] (1-patched/unpatched)
---- | -----
heterogeneous (lots of float with an int at the end; worst-case) | -1.5%
float | 48%
bounded int (magnitude smaller than 2^32) | 48.4%
latin string (all characters in [0,255]) | 32.7%
general int (reasonably uncommon?) | 17.2%
general string (reasonably uncommon?) | 9.2%
tuples of float | 63.2%
tuples of bounded int | 64.8%
tuples of latin string | 55.8%
tuples of general int | 50.3%
tuples of general string | 44.1%
tuples of heterogeneous | 41.5%

### By structure
These are just the benchmarks described in `Objects/listsort.txt`. The first table is the loss we experience if we sort structured heterogeneous lists (worst case: list is already sorted, we go all the way through doing `n` type-checks, and then we only end up doing `n` comparisons. Tragic, but extremely unlikely in practice; in practice, we would usually find the first heterogeneous element early on, and break out of the check, but here, the single, lonely `float` is hiding all the way at the end of the list of `int`, so we don't find it until we've done all `n` checks):

Benchmark (for heterogeneous lists, worst-case) | Percent improvement (1-patched/unpatched)
---- | ----
\sort | -17.2%
/sort | -19.8%
3sort | -18.0%
+sort | -18.8%
%sort | -10.0%
~sort | -2.1%
=sort | -21.3%

The second table is the same benchmark, but on homogeneous lists (int):

Benchmark (for homogeneous lists) | Percent improvement (1-patched/unpatched)
---- | ----
\sort | 54.6%
/sort | 56.5%
3sort | 53.5%
+sort | 55.3%
%sort | 52.4%
~sort | 48.0%
=sort | 45.2%

## Patch summary
Here we describe at a high level what each section of the patch does:

Line numbers in `Objects/listobject.c` | What the lines do
-- | --
1053-1069 |  Define a struct to hold the function pointers we will select in the pre-sort check. This struct then has to be passed in to every function that performs comparison (to keep things in local scope).
1075-1080 | Compare function for heterogeneous lists; just a wrapper for `PyObject_RichCompareBool`. To be selected if all of our pre-checks fail. 
1086-1108 | Compare function for general homogeneous lists; just a wrapper for `ob_type->tp_richcompare`, which is stored by the pre-sort check at `compare_funcs.key_richcompare`. This yields modest optimization (neighbourhood of 10%), but we generally hope we can do better.
1111-1127 | Compare function for lists of *latin* string. During the pre-sort check, we verify that every string in the list uses one character per byte; otherwise, we default to the general homogeneous compare. If this check is even somewhat likely to pass, it's worth it, because the payoff is large, as can be seen in the Benchmarks section. The compare function basically directly accesses the data buffers of the two strings and `memcmp`s them.
1130-1154 | Compare function for lists of *bounded* long. During the pre-sort check, we verify that every int in the list fits in a single machine word. If that check passes, we can use this optimized compare function, which basically directly compares the machine words representing the two ints (taking sign into account). This is faster than the general comparison, which has to figure out which word is most significant for both inputs, etc, in addition to all the type-checking.
1157-1166 | Compare function for lists of float. Doesn't assume anything; just directly compares the two floats, skipping all the unnecessary type-checking. Because `PyFloat_Type->tp_richcompare` does a lot of typechecking that we want to move out of the sort loop, it pays to have this optimized compare available.
1173-1233 | Compare function for lists of *non-empty* tuple. Tuple comparison is optimized on two levels. Namely, after selecting `compare_funcs.key_compare` in the pre-sort check, we run the pre-sort check *again* on the list `T = [x[0] for x in L]` (we don't actually run the check twice, but we do something functionally equivalent to this). If `T` is type-homogeneous, or even better, satisfies the requirements for one of our special-case compares, we can replace the call to `PyObject_RichCompareBool` for the first tuple element with a call to `compare_funcs.tuple_elem_compare`. This allows us to bypass two levels of wasteful safety checks. If the first elements of the two tuples are equal, of course, we have to call `PyObject_RichCompareBool` on subsequent elements; the idea is that this is uncommon in practice.
2168-2212 | First part of the pre-sort check: we set the variables `key_type`, `keys_are_all_same_type`, `ints_are_bounded`, `strings_are_latin`, and `keys_are_in_tuples` (which is 1 iff. every list element is a non-empty tuple, in which case all the other variables refer to the list `[x[0] for x in L]`).
2215-2243 | Second part of the pre-sort check: given values for those variables, select the appropriate compare function. If `keys_are_in_tuples` and `key_type != &PyTuple_Type`, then use the other variables to select `compare_funcs.tuple_elem_compare`, and set `compare_funcs.key_compare = unsafe_tuple_compare`. 

## Selected quotes from the [python-ideas thread](https://mail.python.org/pipermail/python-ideas/2017-March/045106.html)

[Terry Reedy](https://mail.python.org/pipermail/python-ideas/2017-March/045163.html):
>  Do reference this thread, and quote Tim's approval in principle, if he did not post on the tracker.

[Tim Peters](https://mail.python.org/pipermail/python-ideas/2017-March/045145.html):
> Would someone please move the patch along?  I expect it's my fault it's languished so long, since I'm probably the natural person to review it, but I've been buried under other stuff.
> 
> But the patch doesn't change anything about the sorting algorithm itself - even shallow knowledge of how timsort works is irrelevant.  It's just plugging in a different bottom-level object comparison _function_ when that appears valuable.
> 
> I've said from the start that it's obvious (to me ;-) ) that it's an excellent tradeoff.  At worst it adds one simple (pre)pass over the list doing C-level pointer equality comparisons.  That's cheap.  The worst-case
> damage is obviously small, the best-case gain is obviously large, and the best cases are almost certainly far more common than the worst cases in most code.
 
*Later in that message, Tim also pointed out a bug, which has been fixed in this version of the patch.*

<!-- issue-number: bpo-28685 -->
https://bugs.python.org/issue28685
<!-- /issue-number -->

embg committed:
Added data-aware optimizations to list.sort()

embg committed:
Removed trailing whitespace from listsort_impl

embg committed:
fixed typo

embg committed:
Added myself to Misc/ACKS

embg committed:
Made ISLT comment more in line with the current text

embg committed:
Remove newline

embg committed:
untabify

embg committed:
removed newline

embg committed:
simplified description of the tuple compare

embg committed:
grammar

embg committed:
Bugfix -- gcc ignored the error, but clang caught it!

embg committed:
Bugfix -- assertion in unsafe_long_compare was phrased incorrectly.

tonnydourado commented on Objects/listobject.c:
```diff
@@ -1044,6 +1045,193 @@ sortslice_advance(sortslice *slice, Py_ssize_t n)
            if (k)
 
+/* Here we define custom comparison functions to optimize for the cases one commonly 
+ * in practice: homogeneous lists, often of one of the basic types. */
```
Isn't there a "encounters" or "finds" missing here, between "commonly" and "in practice"?

embg replied:
yup, thanks for pointing this out.

embg committed:
fix typo

embg committed:
fix typo

tim-one reviewed (changes requested):
C is a "call by value" language.  In particular, when you pass a struct, the language has to act as if a physical copy of the struct is passed.  The easiest way is for the implementation to actually make a copy every time a struct is passed; while a smart compiler can arrange to do it more efficiently in most cases, it's poor practice to _rely_ on that:

http://www.cs.utsa.edu/~wagner/CS2213/structs/structs_pass.html

That's why you almost never see non-trivial structs passed in C.  Instead a pointer is passed.

The uses of the MergeState struct in the existing sort code are typical examples of that.  Indeed, I believe we'd be better off if you added the new comparison-function members directly _to_ the MergeState struct, and pass around pointers to `ms` (the one MergeState instance) instead.  The top layers of functions already pass pointers to that struct around, so their signatures wouldn't need to change at all.

embg commented:
@tim-one I'm aware that the struct would be copied; my reasoning for wanting to pass by value was [rather nonsensical](https://stackoverflow.com/questions/42659376/define-a-struct-that-contains-a-function-pointer-to-a-function-that-takes-the-st/42659527?noredirect=1). I realize now that I was completely wrong.

The reason I chose to define a new struct instead of adding to `MergeState` was that `gallop_left` and `gallop_right` don't take in `MergeState`, and I thought it would be confusing to add it as a parameter. Your call: should I add `MergeState` to the `gallop_*` signatures, or just change `CompareFuncs` to `CompareFuncs*`? The former be much less invasive, actually.

Last thing: since I've got you here, can I ask you another question? Namely, this entire project was based on the following quote from the documentation:

> Like its identity, an object’s type is also unchangeable. [1]
> 
> [1] It is possible in some cases to change an object’s type, under certain controlled conditions. It generally isn’t a good idea though, since it can lead to some very strange behavior if it is handled incorrectly.

So I figured I can assume the types will stay the same. I mean, what's the point of checking the types if they're going to change on you? However, ppperry supplied the the following counterexample on the tracker:

```
        @functools.total_ordering
	class ClassAssignmentCanBreakChecks():
		def __init__(self, i):
			self._i = i
		def __lt__ (self, other):
			last.__class__ = OrdinaryOldInteger
			return self._i < (other._i if hasattr(other, '_i') else other)
	@functools.total_ordering
	class OrdinaryOldInteger:
		def __init__(self, i):
			self._i = i
		def __lt__(self, other):
			return self._i < (other._i if hasattr(other, '_i') else other)
	lst = [ClassAssignmentCanBreakChecks(i) for i in range(10)]
	random.shuffle(lst)
	last = lst[-1]
	lst.sort()
```

My code would work correctly here, since the compare functions are equivalent sans side effects, but in general something like this could break `unsafe_object_compare`. Do you think this is something worth worrying about? 

Thank you so much for reviewing my code!

tim-one commented:
Yes, I believe it would be best to add the new stuff to the MergeState struct.

Why that's not confusing:  conceptually, we're building an object to do the sort.  All the sort-related helper functions are really methods of that object, and the MergeState struct is that object's private data.  Indeed, that's why the functions that _do_ currently take a MergeState argument list it first:  in analogy with a Python method taking `self` as its first argument.  That some (conceptual) methods do not currently take a MergeState argument first is just an accidental micro-optimization.  Now that some need access to the (conceptual) object's private data, the _obvious_ way to do that is to give them an initial MergeState argument.

If CPython were coded in C++ instead, it really would be obvious ;-)

About the `__class__`-changing example, I don't care provided it doesn't blow up.  But someone else should weigh in on that.  I hope we can get, e.g., Raymond Hettinger to stare at this issue too.

embg commented:
@tim-one Now that you explained it, I'm totally on board; I'll rewrite everything to use `MergeState`. 

About the `__class__`-changing example: there's no way it could segfault if your classes are defined using pure-Python, because the pure-Python operators/functions will raise exceptions if they get bad types (correct me if I'm wrong). So you would just get an exception, which is what is appropriate here anyway IMO. However, if your comparison function were defined using the C API, I'm pretty sure you could set it up to get a segfault.

tim-one commented:
I'm curious about what the patched Python prints for this program:

    def mkn():
        return float("nan")
    n = mkn()
    xs = [(n, i) for i in range(4)]
    ys = [(mkn(), i) for i in range(4)]
    print(sorted(reversed(xs)))
    print(sorted(reversed(ys)))

What it prints now:

    [(nan, 0), (nan, 1), (nan, 2), (nan, 3)]
    [(nan, 3), (nan, 2), (nan, 1), (nan, 0)]

embg commented:
@tim-one Just ran it; it printed this:

```
[(nan, 0), (nan, 1), (nan, 2), (nan, 3)]
[(nan, 3), (nan, 2), (nan, 1), (nan, 0)]
```

embg commented:
@tim-one  Actually, I'm getting weirded out here.

First of all. The test you just gave *should not have worked*. What you were trying to get at is that `PyObject_RichCompareBool` compares equal if references are equal, while the `nan`s don't compare equal. Right? Well, `unsafe_tuple_compare`, which calls `unsafe_float_compare`, doesn't look at the addresses -- it just compares the floats. So it should have given a different answer. I even made it print out which compares it was using, just to be sure, and it was using those two. So why did it work?

Another example. ppperry proposed the following test:
```
class A(int):
	def __lt__(self, other):
		print("zebra")
		A.__lt__ = A.__false_lt__
		return int.__lt__(self, other)
	__true_lt__ = __lt__
	def __false_lt__(self, other):
		print("gizmo")
		A.__lt__ = A.__true_lt__
		return int.__lt__(self, other)
```
The idea is that if you cache the richcompare, you're going to mess up the print statements. But my patched interpreter printed the same as the current implementation!

Finally, I decided to run the original counterexample ppperry proposed, putting in print statements to see what's going on:
```
from random import *
import functools

@functools.total_ordering
class ClassAssignmentCanBreakChecks():
    def __init__(self, i):
        self._i = i
    def __lt__ (self, other):
        print('gizmo')
        last.__class__ = OrdinaryOldInteger
        return self._i < (other._i if hasattr(other, '_i') else other)

    
@functools.total_ordering
class OrdinaryOldInteger:
    def __init__(self, i):
        self._i = i
    def __lt__(self, other):
        print('rocket')
        return self._i < (other._i if hasattr(other, '_i') else other)
lst = [ClassAssignmentCanBreakChecks(i) for i in range(10)]
shuffle(lst)
last = lst[-1]
lst.sort()
```
Again, it matched up with the current implementation!

So my question is: *why is my patch working when it shouldn't be working*??? I have no idea; can you figure this out?

The only explanation I can think of is that maybe heap types all use the same `ob_type->tp_richcompare`, which looks up the appropriate `__lt__` method. Is this correct?

tim-one commented:
Slamming 3 distinct mysteries into one message doesn't really help ;-)

I'll address the one I started:  I expect you'd get different results (in the NaN test) if your tuple compare loop:

    for (i = 0; i < vlen && i < wlen; i++) {

started at `i = 1` instead.  At this point you already checked `v[0]` against `w[0]` twice, and since the "unsafe" float compare said "nope, not less than" both times you decided they must be equal.  But the loop goes on to compare them a third time, using `PyObject_RichCompareBool` which _does_ special-case object identity.  So the program I gave yields the same results with or without the patch.

Which is a really weird "reason" to keep that loop starting at `i = 0`.  Which I didn't realize it did at first, instead believing the _comment_:

    We need to look at v[1:] and w[1:]

embg commented:
@tim-one  Ah! That makes sense. What happened was, I figured, if you're already at the loop, it's going to be slow anyway, so I may as well *exactly* copy-paste from the current implementation to reduce the probability of error. Turns out that saved me! Anyway, your call: should I change it to start at 1 (adding a pointer compare to make it safe)?

As for the other mysteries: I'm convinced it must just be that `ob_type->tp_richcompare` is just a lookup method for `__lt__`, so when the type changes it automatically switches the compare. 

So, basically, the patch has withstood all attempts to break it! Awesome!

embg committed:
Added if (v == w) return 1; to all compares

embg committed:
Merge branch 'fastsort' of https://github.com/embg/cpython into fastsort

embg committed:
Added if (v == w) return 0; to all compares, apologies for previous commit

embg committed:
Added if (v == w) return 0; to all compares, apologies for previous commit

embg committed:
Folded CompareFuncs into MergeState and added safety check to unsafe_object_compare

embg committed:
Folded CompareFuncs into MergeState and added safety check to unsafe_object_compare

embg committed:
formatting

embg committed:
formatting

embg committed:
formatting

embg committed:
formatting

embg committed:
don't need (v==w) for ints/strings

embg committed:
don't need (v==w) for ints/strings

embg commented:
OK, I believe I've addressed all the issues that have been raised. Specifically:

- I got rid of `CompareFuncs` and put the function pointers in `MergeState`.
- I added a safety check to `unsafe_object_compare` that actually makes it pretty safe. Specifically, it checks that it has the correct compare function before comparing, to avoid issues with heap types changing their `__lt__` methods or whatever during the sort. 
- I added `(v==w)` checks, where appropriate, to make it compatible with the current semantics.

embg committed:
went back to i=0 for tuples; we can't infer == from < and >, even with v==w

embg committed:
went back to i=0 for tuples; we can't infer == from < and >, even with v==w

tim-one commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,177 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+    PyObject* res = (*(ms->key_richcompare))(v, w, Py_LT);
+    if (res == NULL)
+        return -1;
+    int ok;
```
I'm not sure what the current coding standards are, but back in my day ;-) all variables had to be declared at the top of the block using them, because strict C required it.  That is, move `int ok;` to the top of the function?

embg replied:
haha, a bit before my time! fixed.

tim-one commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,177 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+    if (res == NULL)
+        return -1;
+    int ok;
+    if (PyBool_Check(res)){
```
Right before this line, should add

    if (res == Py_NotImplemented) {
        Py_DECREF(res);
        return PyObject_RichCompareBool(v, w, Py_LT);
    }

That should repair the latest example on the bug tracker.

embg replied:
thanks, fixed

tim-one reviewed (changes requested):
The patch is making great progress, and "feels close" - nice work!

One thing that needs to be addressed:  tests.  To the extent possible, tests should exercise every line of the new code.

They needn't be fancy - we already know sorting works.  But we do need:

1. Quick sanity checks that all the new special cases actually do sort.

2. Tests that exercise all the counterexamples that turned up during development.  By their very nature, those are cases that weren't screamingly obvious at first, and as a matter of decades of observation those are also cases that are likely to break again when someone later modifies the code ("why not?" won't be screamingly obvious to them either).

embg committed:
move all declarations to top of their blocks

embg committed:
move all declarations to top of their blocks

embg commented:
@tim-one For the tests -- do you want me to commit to the test suite, or just write a script that we can run for review? Here's what I've been using thus far:

```
def check_sorted(L):
    for i in range(len(L)-1):
        assert(L[i+1] >= L[i])

lists = []

lists.append(['z','ab','cad','_ag'])
lists.append(['z','ab','cad','_ag','\uffff'])
lists.append([1,-1,5])
lists.append([1,-1,5,2**60,-2**60])
lists.append([1.1,-1.1,0])
lists.append([1,1.1,5,-2,0.5])
lists.append([b'sdf',b'afd',b'153d'])
lists.append([(),(1,),(2,)])

for l in lists.copy():
    lists.append([(x,1.1,-4,'a') for x in l])
    lists.append([((x,1.1,-4,'a'),42) for x in l])

for l in lists:
    check_sorted(sorted(l))

try:
    sorted(['a',1])
except TypeError:
    pass
```
It's crude, I know. Should I just extend it and post a new one here?

embg committed:
typo

embg committed:
typo

embg committed:
typo

embg committed:
typo

embg committed:
added Py_NotImplemented check to unsafe_object_compare

embg committed:
added Py_NotImplemented check to unsafe_object_compare

tim-one commented:
Tests have to be in the test suite, else they're not really tests ;-)  As the decades go by, the test suite - which gets run all the time (whether people want it to or not) - is the only thing that ensures code continues to work as intended.

embg commented:
@tim-one OK, I just wanted to make sure I knew what you meant. Ya, this should be fun; I've never used `unittest` before! I'll commit them ASAP. Should I just add them to the existing sort test, or make a new file? I don't want to clutter anything up :)

tim-one commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+    assert(v->ob_type == w->ob_type &&
+           v->ob_type->tp_richcompare != NULL);
+  #endif
+    int ok; 
```
In a debug build, the declaration of `ok` still isn't at the top of the block.  If you look at virtually any of CPython's implementation, you'll see that declarations immediately follow a block's opening `{`'

embg replied:
Right, sorry! Fixed (won't show up until I file a new PR, though)

serhiy-storchaka replied:
Current style guide allows declarations not at the start of the block.

tim-one commented:
Not a new file - just add stuff to `Lib/test/test_sort.py`.  It's easy the tenth time you do it ;-)

embg committed:
ACTUALLY moved declarations to the tops of blocks

embg committed:
ACTUALLY moved declarations to the tops of blocks

embg committed:
Merge branch 'fastsort' of https://github.com/embg/cpython into fastsort

embg committed:
Made tuple compare precisely consistent with PyObject_RichCompareBool; it was not earler

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+    if (res == NULL)
+        return -1;
+
+    if (PyBool_Check(res)){
```
Add a space before `{`.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+  #endif
+    int ok; 
+    
+    if (v == w) return 0;
```
Split the line before `return`.

embg replied:
This was actually removed -- it turns out you only need to look at addresses when you're doing equality testing.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+
+/* Latin string compare: safe for any two latin (one byte per char) strings. */
+static int
+unsafe_latin_compare(PyObject* v, PyObject* w, MergeState* ms){
```
Move `{` to separate line.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1044,6 +1045,75 @@ sortslice_advance(sortslice *slice, Py_ssize_t n)
+    /* This function is used by unsafe_object_compare to optimize comparisons
+     * when we know our list is type-homogeneous but we can't assume anything else.
+     * In the pre-sort check it is set equal to key->ob_type->tp_richcompare */
+    PyObject* (*key_richcompare)(PyObject*, PyObject*, int);
```
It is more common to use a space before `*` rather than after it.

The same is everywhere in the new code.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+
+/* Float compare: compare any two floats. */
+static int
+unsafe_float_compare(PyObject *v, PyObject *w, MergeState* ms){
```
Move `{` to separate line.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+    assert(v->ob_type == w->ob_type &&
+           v->ob_type == &PyFloat_Type);
+  #endif
+    if (v == w) return 0;
```
Split the line before `return`.

Or write
```
    return v != w && PyFloat_AS_DOUBLE(v) < PyFloat_AS_DOUBLE(w);
```

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+    
+    int k;
+    
+    if (v == w) return 0;
```
Split the line before `return`.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+
+    /* Is v[0] < w[0]? */
+    k = (*(ms->tuple_elem_compare))(vt->ob_item[0], wt->ob_item[0], ms);
+    if (k < 0)
```
These two checks can be merged:
```
if (k)
    return k;
```
But add a comment.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
+     * set ms appropriately. */
+    if (saved_ob_size > 1) {
+        /* Assume the first element is representative of the whole list. */
+        int keys_are_in_tuples = (lo.keys[0]->ob_type == &PyTuple_Type &&
```
Why not `PyTuple_CheckExact()`?

embg replied:
I could do that. I just wanted it to be consistent with the rest of the code, which also uses `ob_type` compares. But I think that would probably be closer to the current implementation; at least, `PyTuple_RichCompare` (or whatever it's called) does that (IIRC). So I can make this change.

embg replied:
Actually, I just looked up the definition, and I don't think replacing the current code makes any sense:

```
#define PyTuple_CheckExact(op) (Py_TYPE(op) == &PyTuple_Type)
```

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
+        for (i=0; i< saved_ob_size; i++) {
+
+            if (keys_are_in_tuples &&
+                (lo.keys[i]->ob_type != &PyTuple_Type || Py_SIZE(lo.keys[i]) == 0)) {
```
Why not `PyTuple_CheckExact()`?

I would write the second part of the condition as a negation: `!(... && ...)`.

embg replied:
OK, rewritten (second point). For your first point, see my previous comment; I think it makes more sense to keep things consistent.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+static int
+safe_object_compare(PyObject* v, PyObject* w, MergeState* ms)
+{
+  /* No assumptions necessary! */
```
Bad indentation.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+unsafe_object_compare(PyObject* v, PyObject* w, MergeState* ms)
+{
+  /* Modified from Objects/object.c:PyObject_RichCompareBool, assuming: */
+  #ifdef Py_DEBUG
```
Is this `#ifdef` needed? `assert()` is no-op in non-debug build.

embg replied:
Oh, I didn't know that, thanks! Fixed. (in next commit, haven't pushed yet). Same goes for all your other comments.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+{
+  /* Modified from Objects/object.c:PyObject_RichCompareBool, assuming: */
+  #ifdef Py_DEBUG
+    assert(v->ob_type == w->ob_type &&
```
It would be more helpful to use two separate asserts. In case of the failure we could know what condition is false.

serhiy-storchaka replied:
Does `list.sort()` sorts in-place? Can the content of the list be modified during sorting? If this is true, then the condition `v->ob_type == w->ob_type` can be false.

embg replied:
Yes, you are correct. `ppperry` gave an explicit counterexample to this. It's OK, though, because I added the following:
```
    if (v->ob_type->tp_richcompare != ms->key_richcompare)
        return PyObject_RichCompareBool(v, w, Py_LT);
```
However, I should get rid of the `ob_type` assertion: it could legitimately fail.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+  /* Modified from Objects/object.c:PyObject_RichCompareBool, assuming: */
+  #ifdef Py_DEBUG
+    assert(v->ob_type == w->ob_type &&
+           v->ob_type->tp_richcompare != NULL);
```
Is it possible that `v->ob_type` is changed during sorting? If this is true, than the condition `v->ob_type->tp_richcompare != NULL` can be false.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+    
+    if (res == Py_NotImplemented) {
+        Py_DECREF(res);
+        return PyObject_RichCompareBool(v, w, Py_LT);
```
This calls `v->ob_type->tp_richcompare()` the second time.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+    wlen = Py_SIZE(wt);
+
+    /* Well, are either of the tuples are singleton? */
+    if (vlen == 1 || wlen == 1)
```
Counterexample: `(1,) < (1, 2)`.

embg replied:
Yup, I rewrote `unsafe_tuple_compare`, I just think it doesn't update here because the PR is closed. I made it more consistent with the current implementation; specifically, it now does equality testing before less-than testing. The performance gain is smaller, but it was the only way to make it 100% consistent.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+    if (v == w) return 0;
+
+    /* Is v[0] < w[0]? */
+    k = (*(ms->tuple_elem_compare))(vt->ob_item[0], wt->ob_item[0], ms);
```
Since specialized comparison is not recursive, wouldn't be safe to pass `NULL` rather of `ms`? Just for sure that this parameter is not misused.

embg replied:
No. If the compare function is `unsafe_object_compare`, then it could need the `ms` struct. For example: `[(2**60,), (-2**60,)]`. `unsafe_long_compare` doesn't apply because the ints are too big.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+
+    /* Out of options: v[0] == w[0]! We need to look at v[1:] and w[1:].
+     * We can use code copied straight from tupleobject.c:tuplerichcompare: */
+    for (i = 0; i < vlen && i < wlen; i++) {
```
`i = 1;`

embg replied:
Ya, I actually rewrote `unsafe_tuple_compare` to move the less-than after the equality testing, to make sure it's 100% consistent. The equality testing is not optimized; I figured it would be too much code for too little benefit.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
+
+        /* Prove that assumption by checking every key. */
+        int i;
+        for (i=0; i< saved_ob_size; i++) {
```
Normalize spaces.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
+     * How much optimization is safe? We test for homogeneity with respect to 
+     * several properties that are expensive to check at compare-time, and 
+     * set ms appropriately. */
+    if (saved_ob_size > 1) {
```
Shouldn't we just return if `saved_ob_size <= 1`?

Is there a befit from the optimization for `saved_ob_size == 2`? For `saved_ob_size == 3`?

embg replied:
For `<= 1`, I just didn't want to mess with anything. For `2,3`, I just didn't want to have to write additional code to set `ms` appropriately. I think it makes more sense to just have one block of code, instead of a separate block at the end that sets `ms->key_compare = safe_object_compare` for `== 2,3`. My benchmarking indicates that the gain would be negligible, and not worth an extra 3-4 lines...

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
+                                  Py_SIZE(lo.keys[0]) > 0);
+
+        PyTypeObject* key_type = (keys_are_in_tuples ?
+                                  PyTuple_GET_ITEM(lo.keys[0],0)->ob_type :
```
Add a space after `,`.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
+             * lo.keys[i], not lo.keys[i] itself! We verify type-homogeneity 
+             * for lists of tuples in the if-statement directly above. */
+            PyObject* key = (keys_are_in_tuples ?
+                             PyTuple_GET_ITEM(lo.keys[i],0) :
```
Add a space after `,`.

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
+                break;
+            }
+
+            else if (key_type == &PyLong_Type && ints_are_bounded &&
```
`else` is not needed since an alternate branch is ended by `break`.

If `ints_are_bounded` or  `Py_ABS(Py_SIZE(key)) > 1` are false, `key_type == &PyUnicode_Type` still is checked.

I would write:
```
if (key_type == &PyLong_Type) {
    if (ints_are_bounded && Py_ABS(Py_SIZE(key)) > 1)
        ints_are_bounded = 0;
}
else if (key_type == &PyUnicode_Type ...
```

embg replied:
OK!

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
+        /* Choose the best compare, given what we now know about the keys. */
+        if (keys_are_all_same_type) {
+
+            if (key_type == &PyUnicode_Type && strings_are_latin)
```
Since in any case you use empty lines for readability, it would be better to add braces.

embg replied:
Sure

serhiy-storchaka commented on Objects/listobject.c:
```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
+            else if ((ms.key_richcompare = key_type->tp_richcompare) != NULL)
+                ms.key_compare = unsafe_object_compare;
+
+        } else {
```
`else` should be at the start of new line.

serhiy-storchaka commented:
Sorry, I missed that this PR is closed. Is the proposition rejected?

embg commented:
@serhiy-storchaka  Thanks so much for your review! Tim asked me to write tests in `Lib/tests/sort_test.py` and then closed it; I think he wants me to open a new PR once I've written the tests?

tim-one commented:
Ack!  This is the first time I've used github, and I didn't _intend_ to close it.  I must have hit a wrong button :-(  Here, I'll try to open it again.  Sorry!

embg commented:
@tim-one Haha no problem, it's going to be a couple days till my tests are up to snuff :)

embg committed:
Added tests

embg committed:
Implemented all of serhiy-storchaka's changes

embg committed:
Removed braces at the end of unsafe_tuple_compare

embg committed:
Fixed test_safe_object_compare

Comparing bytes and strings yields a warning, not an error, so assertRaises fails.

embg commented:
@serhiy-storchaka Don't know if you saw this or not; I fixed all the issues you raised. As far as I can tell, the patch is 100% finished.

pppery commented on Lib/test/test_sort.py:
The test on line 325 doesn't actually work; the first sort of the non-tupled list already did the `__class__` modification, so L now contains a WackyList2 already and thus the sort would fail even if there was a bug that made sorting one-tuples not handle mid-sort tp_richcompare changes.

rhettinger committed:
Fix spacing around PyObject *

rhettinger committed:
Merge in changes to master

rhettinger committed:
Add news blurb

rhettinger committed:
Update listsort.txt for the optimization

rhettinger committed:
Fix whitespace

rhettinger committed:
Clean-up and fix tests for mutation of __class__.

* Use the with-statement form of self.assertRaises.

* Make the two assertions independent of one another.
  The second test was invalid because the first list
  was already sorted, making the ValueError inevitable.
